### ISSUE TITLE
New Rule: no-else-return

### ISSUE THREAD

**Issue by pertrai1** [pertrai1/eslint-plugin-llm-core#113]
## Problem

LLMs reflexively write `else` blocks after `return` statements. The pattern looks like:

```ts
if (condition) {
  return valueA;
} else {
  return valueB;
}
```

The `else` is provably redundant — if the `if` branch returns, execution never reaches the `else` branch except through the implicit path. LLMs write this pattern because it mirrors how they "think" about branches (case A vs case B), but it adds visual noise and an indentation level that makes the function harder to read.

More critically, this pattern signals the LLM is *not* thinking in guard clauses. If it writes `if/return/else`, it's likely to continue nesting instead of flattening. Catching `else-after-return` nudges the LLM toward the horizontal coding style that reduces overall complexity.

ESLint's built-in `no-else-return` catches the syntax, but its message is minimal: `"Unexpected 'else' after 'return'."` No explanation of why, no guidance on the broader guard clause pattern.

## Rule

Disallow `else` blocks when the preceding `if` branch (or all preceding branches in an `if/else if` chain) contain a `return`, `throw`, `continue`, or `break` statement.

## Examples: Incorrect

```ts
// else after return
function getStatus(code: number): string {
  if (code >= 200 && code < 300) {
    return "ok";
  } else {
    return "error";
  }
}
```

```ts
// else after throw
function assertPositive(value: number): void {
  if (value <= 0) {
    throw new RangeError("Value must be positive");
  } else {
    proceed(value);
  }
}
```

```ts
// else if chain where all paths terminate
function classify(n: number): string {
  if (n > 0) {
    return "positive";
  } else if (n < 0) {
    return "negative";
  } else {
    return "zero";
  }
}
```

```ts
// else after return in nested scope
function process(input: unknown): string {
  if (typeof input === "string") {
    return input.toUpperCase();
  } else {
    return String(input);
  }
}
```

## Examples: Correct

```ts
// Guard clause — no else needed
function getStatus(code: number): string {
  if (code >= 200 && code < 300) {
    return "ok";
  }
  return "error";
}
```

```ts
// Early throw — no else needed
function assertPositive(value: number): void {
  if (value <= 0) {
    throw new RangeError("Value must be positive");
  }
  proceed(value);
}
```

```ts
// if/else if without terminal else — each branch returns
function classify(n: number): string {
  if (n > 0) return "positive";
  if (n < 0) return "negative";
  return "zero";
}
```

```ts
// else where if branch does NOT always terminate — this is fine
function getValue(input: unknown): string {
  if (typeof input === "string" && input.length > 0) {
    console.log("got string");
    // no return — fallthrough is possible
  } else {
    return "default";
  }
  return input;
}
```

## Why it matters

When an LLM writes `else` after `return`, it's treating each branch as a co-equal path rather than using guard clauses to eliminate special cases early. This "diagonal" thinking cascades — the next condition gets nested inside the `else`, and the function grows vertically instead of horizontally.

Forcing removal of the `else` after `return` is the single most effective structural nudge toward flat code. It makes the rest of the function the implicit else case, removing an indentation level for all subsequent logic.

This rule complements `prefer-early-return` (which catches the "entire function body in one if" case) by catching the more granular "unnecessary else after terminal statement" case.

## Scope boundaries

**In scope (flag):**
- `else` blocks where the corresponding `if` (or all `if/else if` branches) contain an unconditional `return`, `throw`, `continue`, or `break`
- `else if` blocks that follow a terminated `if` — suggest removing the `else` keyword
- Both block and non-block consequents (e.g., `if (x) return; else doSomething()`)

**Out of scope (do NOT flag):**
- `else` blocks where the `if` branch does NOT unconditionally terminate (the `else` is meaningful)
- `else if` treated as a flat chain — `if (a) return x; else if (b) return y; else return z;` should flag the outer else but not force removal of the `else if` keywords (this matches ESLint's `no-else-return` with `allowElseIf: false`)
- Ternary expressions — `a ? b : c` is not an `else` block

**False-positive boundaries:**
- Must verify ALL code paths in the `if` body terminate, not just that a `return` exists somewhere. A function call named `returnSomething()` should not be treated as a return statement.
- Must handle `if` bodies with nested control flow that may or may not terminate on all paths. If the `if` body has a `return` inside a nested `if`, the outer `if` does NOT always return.

## Proposed config placement

**`recommended`** — This is a zero-controversial structural improvement. No legitimate code pattern requires `else` after `return`. Every codebase benefits.

Also added to **`style`** config (alongside `prefer-early-return`, `no-redundant-logic`) since it enforces a coding style that reduces nesting.

## Framework-agnostic proof

This rule checks a pure language-level pattern (control flow termination followed by an `else` keyword). It applies identically to any TypeScript/JavaScript code regardless of framework, runtime, or architecture.

## Prior art

- **ESLint built-in `no-else-return`** — Detects the same pattern. This rule differs by providing teaching messages with the plugin's structured format. Default should be `{ allowElseIf: false }` to enforce maximum flatness.
- **`prefer-early-return`** (this plugin) — Catches the broader pattern of "function body wrapped in a single if." Does NOT catch `if/return/else` inside a function — that's a different (and more common) case.
- **`max-nesting-depth`** (this plugin) — Measures overall nesting. An unnecessary `else` adds +1 nesting for the rest of the function. This rule addresses the *cause*, `max-nesting-depth` measures the *symptom*.
- **`no-redundant-logic`** (this plugin) — Already has an `unnecessaryElse` message ID that catches a **subset** of this pattern: `else` after `return`/`throw` where the alternate is a simple single statement. See **Overlap with `no-redundant-logic`** section below.

## Options

```ts
type Options = [
  {
    allowElseIf?: boolean; // Default: false (stricter than ESLint's default of true)
  },
];
```

Setting `allowElseIf: true` would permit `else if` after a return but still flag terminal `else`. The default of `false` enforces maximum horizontal style — convert all `else if` chains to sequential guard clauses.

---

## Overlap with `no-redundant-logic`

**Important:** The existing `no-redundant-logic` rule (Pattern 2: `checkUnnecessaryElse`) already detects a subset of this pattern:

| Aspect | `no-redundant-logic` (existing) | `no-else-return` (proposed) |
|---|---|---|
| What it catches | `else` after `return`/`throw` where the alternate is a **single** statement | `else` after `return`/`throw`/`continue`/`break` with **any** alternate body |
| Terminal statements | `return`, `throw` only | `return`, `throw`, `continue`, `break` |
| `else if` chains | Skipped (`if (node.alternate.type === IfStatement) return`) | Flagged when `allowElseIf: false` |
| Alternate body size | Must be simple (single return/throw) | Any size (complex else blocks too) |
| Auto-fix | Yes (suggests removal) | TBD — likely suggestions, not auto-fix |
| Message focus | "Unnecessary else — the preceding if block always exits" | Teaching message about guard clause pattern and horizontal coding |

### Resolution options

1. **Expand `no-redundant-logic`** — Add `continue`/`break` to its terminal statement check, expand to handle non-simple alternates and `else if` chains, and update the message. Risk: `no-redundant-logic` becomes a catch-all, diluting its "redundant logic" identity.

2. **New rule as described, mark the overlap** — Ship `no-else-return` with full coverage. Have `no-redundant-logic` defer to `no-else-return` when both would fire on the same node (check if the new rule is enabled and skip). Clean separation of concerns.

3. **New rule, retire the overlap in `no-redundant-logic`** — Remove the `unnecessaryElse` pattern from `no-redundant-logic` entirely and let `no-else-return` own this space. Breaking change for users who only have `no-redundant-logic` enabled.

**Recommendation:** Option 2 — keeps both rules independently correct, avoids a breaking change, and the overlap only fires when both rules are enabled (which is the `recommended` config). The messages are different enough that double-reporting is the real concern, not the detection itself.

---

## Proposed Error Message

For discussion — the message the LLM receives when this rule fires. This needs to be distinct from the existing `no-redundant-logic` message and should teach the broader guard clause pattern.

### Existing message in `no-redundant-logic` (for comparison)

```
Unnecessary `else` — the preceding `if` block always exits.

Why: When an `if` block ends with `return` or `throw`, the `else` branch
is unreachable via fall-through. Keeping the `else` adds nesting without
benefit and obscures the control flow.

How to fix:
  Before: if (status === 'active') {
            return 'Active';
          } else {
            return 'Inactive';
          }
  After:  if (status === 'active') {
            return 'Active';
          }
          return 'Inactive';
```

### Proposed message for `no-else-return`

```
Unnecessary `else` after {{ keyword }} — remove the else block and dedent the remaining code.

Why: When an if branch exits (return, throw, continue, or break), the code below it
is guaranteed to run only when the condition was false. The else keyword and its braces
add indentation without adding meaning. Removing it flattens the code, making the
remaining logic easier to read and modify.

How to fix:
  Remove the else keyword and dedent the block contents.

  Before: if (isValid(data)) {
            return process(data);
          } else {
            logInvalid(data);
            return fallback;
          }
  After:  if (isValid(data)) {
            return process(data);
          }
          logInvalid(data);
          return fallback;

  For else-if chains, convert each branch to a separate guard clause:

  Before: if (status === 'active') {
            return 'Active';
          } else if (status === 'pending') {
            return 'Pending';
          } else {
            return 'Unknown';
          }
  After:  if (status === 'active') return 'Active';
          if (status === 'pending') return 'Pending';
          return 'Unknown';
```

### Design decisions to discuss

1. **`{{ keyword }}` placeholder** — Should the message dynamically say "after `return`", "after `throw`", etc.? This adds specificity but increases message template complexity. Alternative: just say "after a statement that exits the function" and keep it static. The existing `no-redundant-logic` message doesn't distinguish between return and throw.

2. **Two before/after examples** — One for simple else, one for else-if chains. This is more guidance than most rules provide (e.g., `prefer-early-return` has one example). Worth the extra length? The else-if case is the most common LLM pattern, so showing the transformation explicitly may reduce fix iterations.

3. **"exits the function" vs. "exits the block"** — `continue` and `break` don't exit the function, they exit the loop/switch. Should the Why section say "exits the current flow" or enumerate the statements? The current wording says "return, throw, continue, or break" which is explicit but long.

4. **Relationship to `no-redundant-logic`'s message** — If both rules fire on the same node (see overlap section), the user/LLM gets two messages for the same violation. The messages say similar things in different words. Should we add a guard in `no-redundant-logic` to skip when `no-else-return` is enabled? Or accept the redundancy since the messages teach different angles (one: "the else is unreachable"; the other: "remove it to flatten your code")?

5. **Should this rule provide auto-fix suggestions?** — `no-redundant-logic` provides a suggestion fixer for its `unnecessaryElse` case. This rule covers more cases (else-if chains, non-simple alternates), some of which are harder to auto-fix correctly (dedenting a multi-line else block requires tracking indentation). Should we provide suggestions for simple cases only, or skip suggestions entirely to match the plugin's philosophy of "suggestions, not auto-fixes, when semantics could change"?

---

## Definition of Done

- [ ] Rule file: `src/rules/no-else-return.ts` using `createRule`
- [ ] Exported from `src/rules/index.ts` (alphabetical)
- [ ] Added to `styleRules` and `recommendedRules` in `src/index.ts`
- [ ] Teaching message follows the what/why/how format with concrete before/after examples
- [ ] Detects `else` after unconditional `return`, `throw`, `continue`, and `break`
- [ ] Does NOT flag when `if` body does not unconditionally terminate on all paths
- [ ] `allowElseIf` option works (default `false`)
- [ ] Overlap with `no-redundant-logic` resolved (see overlap section)
- [ ] Tests: `tests/rules/no-else-return.test.ts` covering:
  - `else` after `return` (flag)
  - `else` after `throw` (flag)
  - `else` after `continue` in loops (flag)
  - `else` after `break` in switch/loops (flag)
  - `else` where `if` body has a return inside a nested `if` — does NOT always terminate (do NOT flag)
  - `else if` after return with `allowElseIf: false` (flag the `else if`)
  - `else if` after return with `allowElseIf: true` (do NOT flag the `else if`, but DO flag terminal `else`)
  - Non-block consequent: `if (x) return; else doSomething()` (flag)
  - Arrow functions with expression bodies (no false positives)
  - `else` after early return in a function with multiple guard clauses
- [ ] Docs: `docs/rules/no-else-return.md`
- [ ] `npm run update:eslint-docs` run
- [ ] `npm test && npm run lint && npm run build` all pass
- [ ] RED/GREEN/REFACTOR cycle followed — `test:` commits before `feat:` commits
